Add support for pFUnit version 4

amklinv-nnl opened the pull request:
Please see #11642 for more details @tclune @sethrj

amklinv-nnl committed:
pFUnit: Added support for version 4

pFUnit v4 uses submodules, so we must fetch from the repo rather
than grabbing the tarball (see #11642).

amklinv-nnl committed:
pFUnit: Added conflicts

pFUnit 4 causes an internal compiler error with gcc 7.2.0, and
several pFUnit versions are incompatible with shared libraries.

amklinv-nnl committed:
pFUnit: Added conflicts for version 4

Verson 4 uses Fortran 2008 features and cannot be built with gcc
compilers prior to 8.4.

sethrj commented on var/spack/repos/builtin/packages/pfunit/package.py:
```diff
@@ -40,7 +41,13 @@ class Pfunit(CMakePackage):
+    depends_on('m4', when='@4.1.5:4.1.9')
 
+    conflicts("%gcc@7.2.0", when="@4.0.0:",
+              msg='pFUnit v4 cannot be built with gcc 7.2.0. Please choose a different compiler.')
```
Is it just 7.2 or all versions up to and including 7.2? The message repeats the restrictions but not why they're needed. I'd change to something like:
```
conflicts("%gcc@:7.2", when="@4.0.0:",
               msg='Older versions of GCC do not support the Fortran 2008 features required by new pFUnit')
```

sethrj commented on var/spack/repos/builtin/packages/pfunit/package.py:
```diff
@@ -40,7 +41,13 @@ class Pfunit(CMakePackage):
+              msg='pFUnit v4 cannot be built with gcc 7.2.0. Please choose a different compiler.')
+    # See https://github.com/Goddard-Fortran-Ecosystem/pFUnit/pull/179
+    conflicts("+shared", when="@4.0.0:4.1.9")
+    conflicts("%gcc@:8.3.9", when="@4.0.0:")
```
This conflict is redundant with the one on line 46. I'm guessing you're trying to say "it conflicts with GCC < Version('9')" but the spack versioning notation is pretty weird (especially with open/closed ranges). I think `%gcc@:8.999` might be a more standard way to write the restriction.

amklinv-nnl replied:
Good catch!

sethrj commented on var/spack/repos/builtin/packages/pfunit/package.py:
```diff
@@ -40,7 +41,13 @@ class Pfunit(CMakePackage):
 
     depends_on('python@2.7:', type=('build', 'run'))  # python3 too!
     depends_on('mpi', when='+mpi')
+    depends_on('m4', when='@4.1.5:4.1.9')
```
It's likely that since 4.1.9 is the latest, future requirements will also include 4.1.9. I'd change this to `@4.1.5:`.

sethrj reviewed (changes requested):
Take a look at the couple of comments about conflicts and versions ranges.

amklinv-nnl committed:
pFUnit: Fixed conflicts/dependencies as suggested

amklinv-nnl commented:
Thank you for the review @sethrj! I just noticed that @citibeth is listed as the maintainer of this package, so I am tagging her as well.

sethrj reviewed (approved):
Looks good to me, thanks for the updates!

adamjstewart reviewed (approved):
@citibeth can you review?

citibeth commented on var/spack/repos/builtin/packages/pfunit/package.py:
```diff
@@ -14,12 +14,13 @@ class Pfunit(CMakePackage):
-    # version('4.0.0',  sha256='b8b6470f2b1e2b19c164c244c10e803bd69c8da9a6a5a65ba7c479fb8b92a1e1') # noqa: E501
+    version('4.1.9',  tag='v4.1.9', submodules=True)
+    version('4.1.5',  tag='v4.1.5', submodules=True)
+    version('4.0.0',  tag='v4.0.0', submodules=True)
```
Use the `/archive` mechanism to download tarballs as before, instead of git.  This URL works.  Note the 'v' has been added to the git tag starting with version 4.  That should now be the default; whereas earlier versions should get their own special URL.
```
curl -LO https://github.com/Goddard-Fortran-Ecosystem/pFUnit/archive/v4.1.9.tar.gz
```

amklinv-nnl replied:
Thank you for the review!  Can you explain why the archive mechanism is preferred to git?

citibeth commented on var/spack/repos/builtin/packages/pfunit/package.py:
```diff
@@ -14,12 +14,13 @@ class Pfunit(CMakePackage):
 
     homepage = "http://pfunit.sourceforge.net/"
     url      = "https://github.com/Goddard-Fortran-Ecosystem/pFUnit/archive/3.2.9.tar.gz"
```
Change to:
```
url      = "https://github.com/Goddard-Fortran-Ecosystem/pFUnit/archive/v4.1.10.tar.gz"
```

citibeth commented on var/spack/repos/builtin/packages/pfunit/package.py:
```diff
@@ -14,12 +14,13 @@ class Pfunit(CMakePackage):
-    # Currently investigating build fails for v 4.0.0.
-    # See discussion in PR #11642.
-    # version('4.0.0',  sha256='b8b6470f2b1e2b19c164c244c10e803bd69c8da9a6a5a65ba7c479fb8b92a1e1') # noqa: E501
+    version('4.1.9',  tag='v4.1.9', submodules=True)
```
Add v4.1.10 (just 16hr ago it was released)

citibeth commented on var/spack/repos/builtin/packages/pfunit/package.py:
```diff
@@ -14,12 +14,13 @@ class Pfunit(CMakePackage):
+    version('4.1.9',  tag='v4.1.9', submodules=True)
+    version('4.1.5',  tag='v4.1.5', submodules=True)
+    version('4.0.0',  tag='v4.0.0', submodules=True)
     version('3.3.3',  sha256='9f673b58d20ad23148040a100227b4f876458a9d9aee0f0d84a5f0eef209ced5')
```
These older versions will need the `url=` construct because their URLs no longer conform to the current standard.

amklinv-nnl replied:
Does url_for_version not take care of that?  I successfully installed 3.3.3 on my system with this recipe, but it's possible that I accidentally used a cached tarball.

citibeth commented on var/spack/repos/builtin/packages/pfunit/package.py:
```diff
@@ -40,7 +41,12 @@ class Pfunit(CMakePackage):
+    conflicts("%gcc@:8.3.9", when="@4.0.0:", msg='Older versions of GCC do '
+              'not support the Fortran 2008 features required by new pFUnit.')
+    # See https://github.com/Goddard-Fortran-Ecosystem/pFUnit/pull/179
+    conflicts("+shared", when="@4.0.0:4.1.9")
```
Does it also conflict with `@4.1.10`?  If so should problaby do `when="@4.0.0:"`

citibeth commented:
Because it is checksummable.  With the git mechanism, someone could hijack
the pfunit account, add malicious code to it, and Spack would never know
the difference.

On Thu, Jul 30, 2020 at 08:21 Alicia Klinvex <notifications@github.com>
wrote:

> *@amklinv-nnl* commented on this pull request.
> ------------------------------
>
> In var/spack/repos/builtin/packages/pfunit/package.py
> <https://github.com/spack/spack/pull/17683#discussion_r462956348>:
>
> >
>      maintainers = ['citibeth']
>
> -    # Currently investigating build fails for v 4.0.0.
> -    # See discussion in PR #11642.
> -    # version('4.0.0',  sha256='b8b6470f2b1e2b19c164c244c10e803bd69c8da9a6a5a65ba7c479fb8b92a1e1') # noqa: E501
> +    version('4.1.9',  tag='v4.1.9', submodules=True)
> +    version('4.1.5',  tag='v4.1.5', submodules=True)
> +    version('4.0.0',  tag='v4.0.0', submodules=True)
>
> Thank you for the review! Can you explain why the archive mechanism is
> preferred to git?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/spack/spack/pull/17683#discussion_r462956348>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAOVY53ZH7FTZXD7YCMXYZDR6FQU7ANCNFSM4PGDMOMA>
> .
>
-- 
Elizabeth Ashley Fischer, she/her
University of Alaska, Geophysical Institute
cell: 617-308-0436

UA is an AA/EO employer and educational institution and prohibits illegal
discrimination against any individual: www.alaska.edu/nondiscrimination.

tclune commented:
In case it matters ... this thread reminded me that I had not uploaded a tar ball for the latest pFUnit release (v4.1.10) that includes the submodules.

citibeth commented:
If you do, would that change the required URLs?

On Thu, Jul 30, 2020 at 08:30 Tom Clune <notifications@github.com> wrote:

> In case it matters ... this thread reminded me that I had not uploaded a
> tar ball for the latest pFUnit release (v4.1.10) that includes the
> submodules.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/spack/spack/pull/17683#issuecomment-666335157>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAOVY56ZTWPNEZV6LR7TJ6TR6FRVZANCNFSM4PGDMOMA>
> .
>
-- 
Elizabeth Ashley Fischer, she/her
University of Alaska, Geophysical Institute
cell: 617-308-0436

UA is an AA/EO employer and educational institution and prohibits illegal
discrimination against any individual: www.alaska.edu/nondiscrimination.

tclune commented:
The URL for the new upload is:
  https://github.com/Goddard-Fortran-Ecosystem/pFUnit/releases/download/v4.1.10/pFUnit-4.1.10.tar

The URL for the tar autocreated by GitHub (but sans submodules) is:
  https://github.com/Goddard-Fortran-Ecosystem/pFUnit/archive/v4.1.10.tar.gz

amklinv-nnl committed:
pFUnit: Version 4 no longer fetches from git

Checksummable files are fetched instead.

amklinv-nnl commented:
I updated the PR with the requested changes.  Some versions were removed because there were no tarballs available for download.

citibeth commented on var/spack/repos/builtin/packages/pfunit/package.py:
```diff
@@ -40,7 +37,12 @@ class Pfunit(CMakePackage):
 
+    conflicts("%gcc@:8.3.9", when="@4.0.0:", msg='Older versions of GCC do '
+              'not support the Fortran 2008 features required by new pFUnit.')
+    # See https://github.com/Goddard-Fortran-Ecosystem/pFUnit/pull/179
```
Does the comment lin line 44 (PR 179) go with the "gcc@8.3.9" conflicts?  If so, it should be moved just above the "conflicts" statement.

amklinv-nnl replied:
Nope, it goes with the shared conflict.

citibeth commented on var/spack/repos/builtin/packages/pfunit/package.py:
```diff
@@ -40,7 +37,12 @@ class Pfunit(CMakePackage):
+    conflicts("%gcc@:8.3.9", when="@4.0.0:", msg='Older versions of GCC do '
+              'not support the Fortran 2008 features required by new pFUnit.')
+    # See https://github.com/Goddard-Fortran-Ecosystem/pFUnit/pull/179
+#    conflicts("+shared", when="@4.0.0:")
```
Why is this commented out?  Should we add a comment above it explaining why?  Or fully remove it?  Or put it back in?

amklinv-nnl replied:
Good catch.  I didn't mean to leave it commented out.

citibeth commented on var/spack/repos/builtin/packages/pfunit/package.py:
```diff
@@ -50,6 +52,16 @@ def patch(self):
+    def url_for_version(self, version):
+        # Version 4 uses a different URL syntax than previous versions
+        url_base = "https://github.com/Goddard-Fortran-Ecosystem/pFUnit"
+        if int(str(version.up_to(1))) >= 4:
```
@adamjstewart Isn't there a simpler wayto get the first item in a version tuple?

adamjstewart replied:
I would use:
```suggestion
        if version >= Version('4'):
```

citibeth commented:
@tclune See the `patch()` method in this recipe; is there any way pFUnit can be modified upstream to eliminate the need for this patch?

tclune commented:
@citibeth Sorry - I don't know how to look for "this recipe".  Could you send a permalink?  Happy to help.

citibeth commented:
> @citibeth Sorry - I don't know how to look for "this recipe". Could you send a permalink? Happy to help.

See here line 49:

https://github.com/spack/spack/blob/e203361865b6090a0ff4bbd843cf848923e893f9/var/spack/repos/builtin/packages/pfunit/package.py

tclune commented:
OK - what is driving the need for that patch?  -DMPI is not even a current option for pFUnit.  There is a -DSKIP_MPI=YES/NO.

Perhaps if I explain the intent in pFUnit you could then educate me a bit further.   Starting in 4.0, pFUnit detects MPI using ```find_package(MPI ...)```.   If MPI is found, then the MPI extensions are built.  Otherwise they are not.   The extensions are in the form of _additional_ libraries (and includes).  Thus unlike in v3.x, there is no need to have 2 separate builds to support serial and parallel projects.   Rather, a serial project links against 'funit' and a parallel project links against 'pfunit' (and 'funit').

The -DSKIP_MPI option is to force CMake to skip the ```find_package(MPI)``` step and only build the serial interfaces.

amklinv-nnl commented:
@tclune I did not add the patch and can't speak to its intent.  I assume it is unnecessary for version 4 though.  Should I disable that patch for version 4?

amklinv-nnl committed:
pFUnit: Simplify major version check

Co-authored-by: Adam J. Stewart <ajstewart426@gmail.com>

tclune commented:
That would be my recommendation.

amklinv-nnl committed:
pFUnit: Removed unnecessary patch for v4

The patch is still applied to v3.

citibeth commented:
> Perhaps if I explain the intent in pFUnit you could then educate me a bit further. Starting in 4.0, pFUnit detects MPI using `find_package(MPI ...)`. If MPI is found, then the MPI extensions are built. Otherwise they are not. The extensions are in the form of _additional_ libraries (and includes). Thus unlike in v3.x, there is no need to have 2 separate builds to support serial and parallel projects. Rather, a serial project links against 'funit' and a parallel project links against 'pfunit' (and 'funit').

This kind of automagic was common with Autotools, but I believe is considered not best practice with CMake.  It can result in unpredictability.

> The -DSKIP_MPI option is to force CMake to skip the `find_package(MPI)` step and only build the serial interfaces.

Definitely we need to us this in the Spack recipe.

citibeth reviewed (changes requested):
The `-DSKIP_MPI` should be used in the Spack build, so that `pfunit~mpi` NEVER builds an MPI version (even if there's some trashy MPI library sitting around in the system path); and `pfunit+mpi` always builds an MPI version (with the MPI selected by Spack as appropriate).

tclune commented:
What is the best practice in CMake then?    To require the user to specify YES/NO on the use of MPI?  And then with YES, the ```find_package()``` call would include the ```REQUIRED``` keyword?     

I could add the REQUIRED flag for ```-DSKIP_MPI=NO``` and make it optional if ```SKIP_MPI`` is not specified.   Seems like that would satisfy backward compatibility and get close to what you want.   Of course with this change it would be nice to have the more positive ```USE_MPI``` than the negation SKIP being used now.

citibeth commented:
On Tue, Aug 4, 2020 at 11:17 AM Tom Clune <notifications@github.com> wrote:

> What is the best practice in CMake then? To require the user to specify
> YES/NO on the use of MPI? And then with YES, the find_package() call
> would include the REQUIRED keyword?
>

That's correct.  Features should be turned on/off by the user, not
automagically based on what the build can scrounge up.  If the user turns
on a feature and the required dependencies cannot be found, then the build
throws an error.


> I could add the REQUIRED flag for -DSKIP_MPI=NO and make it optional if SKIP_MPI``
> is not specified. Seems like that would satisfy backward compatibility and
> get close to what you want. Of course with this change it would be nice to
> have the more positive USE_MPI``` than the negation SKIP being used now.
>

I agree, I don't like negative logic.  From the Spack point of view, we
need to just use what's there right now.

amklinv-nnl commented:
@tclune FYI, I hit an issue as I was testing my latest changes.

```
5 errors found in build log:
     115
     116      275 |            & values, size(values), MPI_INTEGER, &
     117          |             1
     118    ......
     119      298 |            & values_, size(values), MPI_LOGICAL, &
     120          |             2
  >> 121    Error: Type mismatch between actual argument at (1) and actual argument at (2) (INTEGER(4)/LOGICAL(4)).
     122    /tmp/klinveam/spack-stage/spack-stage-pfunit-3.3.3-rasd4ft5hcapocfcxzm5duytvk75klov/spack-src/source/MpiCon
            text.F90:180:25:
     123
     124      180 |       call mpi_allreduce(value, tmp, 1, MPI_INTEGER, MPI_SUM, &
     125          |                         1
     126    ......
     127      325 |       call MPI_Allreduce(q, anyQ, 1, MPI_LOGICAL, MPI_LOR, &
     128          |                         2
  >> 129    Error: Type mismatch between actual argument at (1) and actual argument at (2) (INTEGER(4)/LOGICAL(4)).
     130    /usr/bin/cmake -E cmake_copy_f90_mod /home/klinveam/spack/opt/spack/linux-ubuntu18.04-skylake/gcc-10.1.0/pf
            unit-3.3.3-rasd4ft5hcapocfcxzm5duytvk75klov/include/throwfundamentaltypes_mod.mod source/CMakeFiles/pfunit.
            dir/throwfundamentaltypes_mod.mod.stamp GNU
     131    source/CMakeFiles/pfunit.dir/build.make:903: recipe for target 'source/CMakeFiles/pfunit.dir/MpiContext.F90
            .o' failed
  >> 132    make[2]: *** [source/CMakeFiles/pfunit.dir/MpiContext.F90.o] Error 1
     133    make[2]: *** Waiting for unfinished jobs....
     134    make[2]: Leaving directory '/tmp/klinveam/spack-stage/spack-stage-pfunit-3.3.3-rasd4ft5hcapocfcxzm5duytvk75
            klov/spack-build'
     135    CMakeFiles/Makefile2:183: recipe for target 'source/CMakeFiles/pfunit.dir/all' failed
  >> 136    make[1]: *** [source/CMakeFiles/pfunit.dir/all] Error 2
     137    make[1]: Leaving directory '/tmp/klinveam/spack-stage/spack-stage-pfunit-3.3.3-rasd4ft5hcapocfcxzm5duytvk75
            klov/spack-build'
     138    Makefile:173: recipe for target 'all' failed
  >> 139    make: *** [all] Error 2
```

It looks like gfortran 10 is a bit more finicky about datatypes than 9 was.

amklinv-nnl committed:
pFUnit: Modified MPI flag for v4

pFUnit v3 and v4 use different CMake flags to enable/disable MPI
support. Also added a conflict for v3 with MPI enabled using
gfortran 10, since newer gfortran is more finicky about datatypes.

citibeth commented on var/spack/repos/builtin/packages/pfunit/package.py:
```diff
@@ -61,10 +74,17 @@ def cmake_args(self):
+                args.append('-DSKIP_MPI=NO')
+            else:
+                args.append('-DMPI=YES')
+            args.extend(['-DMPI_USE_MPIEXEC=YES',
```
Looks good.  My only suggestion would be to reverse the nesting of `spec.satisfies('+mpi')` and `spec.satisfies('@4.0.0:'`.  I think it will be easier to follow the logic for a particular version if the version check is the outer `if` statement.

amklinv-nnl replied:
Sure, I can take care of that right now.  Thank you for the reviews!

amklinv-nnl replied:
Actually, this will complicate the logic of

```
            args.extend(['-DMPI_USE_MPIEXEC=YES',
                         '-DCMAKE_Fortran_COMPILER=%s' % spec['mpi'].mpifc])
```
which happens for +mpi no matter the version.  Would you still like this change?

Edit: I was wrong.  It looks fine.  Please accept the PR if you approve of the changes.

amklinv-nnl committed:
pFUnit: Rearranged mpi logic

tclune commented:
@amklinv-nnl Sorry almost missed your post during the back and forth with @citibeth.

Yes - this is a new strictness in gfortran 10.  I guess I've not yet built with MPI under that version.

The correction is to add "-fallow-argument-mismatch" for version >= 10.   I can try to have a patch by tomorrow morning.

amklinv-nnl commented:
@tclune No rush.  I've added a conflict to the spack package to prevent people from trying to build v3 with mpi using gfortran 10 for now.  Oddly, I didn't see the issue with v4.

tclune commented:
I had reproducing on my end as well.   I found that I could get the error if I used ```include 'mpif.h'``` instead of ```use mpi```.  From this I conclude that it may depend on which flavor/version of MPI is being used.   Depends on how the Fortran wrappers for MPI declare the overloads.   (Easy if they use the F2008 ```TYPE(*)``` feature.)    It may simply be that I'm using a newer version of OpenMPI.

adamjstewart commented on var/spack/repos/builtin/packages/pfunit/package.py:
```diff
@@ -13,13 +13,10 @@ class Pfunit(CMakePackage):
     serial and MPI-parallel software written in Fortran."""
 
     homepage = "http://pfunit.sourceforge.net/"
-    url      = "https://github.com/Goddard-Fortran-Ecosystem/pFUnit/archive/3.2.9.tar.gz"
```
You should keep a url here so that `spack versions` and `spack checksum` still work. You can use the newer URL format.

amklinv-nnl replied:
Thanks for the advice.  This helped me find a new version of pFUnit!  Expect a new commit shortly.

adamjstewart commented on var/spack/repos/builtin/packages/pfunit/package.py:
```diff
@@ -40,16 +37,32 @@ class Pfunit(CMakePackage):
 
     depends_on('python@2.7:', type=('build', 'run'))  # python3 too!
     depends_on('mpi', when='+mpi')
+    depends_on('m4', when='@4.1.5:')
```
This is usually a build dependency:
```suggestion
    depends_on('m4', when='@4.1.5:', type='build')
```

adamjstewart reviewed (commented):
Couple last minute suggestions before this is good to merge.

amklinv-nnl committed:
pFUnit: changed m4 to a build dependency

Co-authored-by: Adam J. Stewart <ajstewart426@gmail.com>

amklinv-nnl committed:
pFUnit: Added URL back

I did not realize it was needed by "spack versions" and
"spack checksum". Thanks @adamjstewart!
